File: 077_Combinations.java
// Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.

// For example,
// If n = 4 and k = 2, a solution is:

// [
//   [2,4],
//   [3,4],
//   [2,3],
//   [1,2],
//   [1,3],
//   [1,4],
// ]

public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> finalAns = new ArrayList();
        int[] tmpAns = new int[k];
        dfs(tmpAns, 0, 1, n, finalAns);
        return finalAns;
    }
    public void dfs(int[] tmpAns, int curIdx, int startNum, int maxNum, List<List<Integer>> finalAns){
        if (curIdx == tmpAns.length){
            List<Integer> newAns = new ArrayList();
            for(int elem: tmpAns){
                newAns.add(elem);
            }
            finalAns.add(newAns);
            return;
        }
        int maxStart = maxNum - tmpAns.length + curIdx + 1;
        for(int i = startNum; i <= maxStart; i++){
            tmpAns[curIdx] = i;
            dfs(tmpAns, curIdx+1, i+1, maxNum, finalAns);
        }
    }
}

File: python/077_Combinations.py
# Given two integers n and k, return all possible
# combinations of k numbers out of 1 ... n.

# For example,
# If n = 4 and k = 2, a solution is:

# [
#   [2,4],
#   [3,4],
#   [2,3],
#   [1,2],
#   [1,3],
#   [1,4],
# ]
# Subscribe to see which companies asked this question


# Runtime: 384ms
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        def helper(ans,used, k, maxnum):
            if k==0:
                res.append(ans)
                return
            if maxnum+k>=n+1:
                return
            for num in xrange(maxnum,n):
                if not used[num+1]:
                    used[num+1]=1
                    helper(ans+[num+1], used, k-1, num)
                    used[num+1]=0
        res = []
        used = {(i+1):0 for i in xrange(n)}
        helper([],used, k, 0)
        return res

if __name__ == '__main__':
    sol = Solution()
    print sol.combine(5,2)
    print sol.combine(1,1)
    print sol.combine(16,12)
